Publishing native AOT fails on Fedora 40, Documentation suggests packages only for Ubuntu

### Describe the bug
I've seen #37995 and the solution doesn't generally apply to Fedora 40. The dotnet documentation is unfortunately Ubuntu specific and would not apply to RHEL distros or Fedora.

Instead you need to install `zlib-ng-devel`, and `zlib-ng-compat-devel`.

### To Reproduce
```
chipc@fedora:~/workspaces/slowgames$ mkdir testproj
chipc@fedora:~/workspaces/slowgames$ cd testproj
chipc@fedora:~/workspaces/slowgames/testproj$ dotnet new console
The template "Console App" was created successfully.

Processing post-creation actions...
Restoring /home/chipc/workspaces/slowgames/testproj/testproj.csproj:
  Determining projects to restore...
  Restored /home/chipc/workspaces/slowgames/testproj/testproj.csproj (in 224 ms).
Restore succeeded.
chipc@fedora:~/workspaces/slowgames/testproj$ dotnet publish -c Release -r linux-x64 -p:PublishAOT=true
MSBuild version 17.8.5+b5265ef37 for .NET
  Determining projects to restore...
  Restored /home/chipc/workspaces/slowgames/testproj/testproj.csproj (in 406 ms).
  testproj -> /home/chipc/workspaces/slowgames/testproj/bin/Release/net8.0/linux-x64/testproj.dll
  Generating native code
  /usr/bin/ld.bfd: cannot find -lz: No such file or directory
collect2 : error : ld returned 1 exit status [/home/chipc/workspaces/slowgames/testproj/testproj.csproj]
/home/chipc/.nuget/packages/microsoft.dotnet.ilcompiler/8.0.5/build/Microsoft.NETCore.Native.targets(366,5): error MSB3073: The command ""gcc" "obj/Release/net8.0/linux-x64/native/testproj.o" -o "bin/Release/net8.0/linux-x64/native/testproj" -Wl,--version-script=obj/Release/net8.0/linux-x64/native/testproj.exports -Wl,--export-dynamic -gz=zlib -fuse-ld=bfd /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/sdk/libbootstrapper.o /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/sdk/libRuntime.WorkstationGC.a /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/sdk/libeventpipe-disabled.a /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/sdk/libstdc++compat.a /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/framework/libSystem.Native.a /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/framework/libSystem.Globalization.Native.a /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/framework/libSystem.IO.Compression.Native.a /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/framework/libSystem.Net.Security.Native.a /home/chipc/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/8.0.5/framework/libSystem.Security.Cryptography.Native.OpenSsl.a -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -pthread -ldl -lz -lrt -lm -pie -Wl,-pie -Wl,-z,relro -Wl,-z,now -Wl,--eh-frame-hdr -Wl,--discard-all -Wl,--gc-sections" exited with code 1. [/home/chipc/workspaces/slowgames/testproj/testproj.csproj]
chipc@fedora:~/workspaces/slowgames/testproj$ sudo dnf install zlib-ng-compat-devel
Last metadata expiration check: 0:54:44 ago on Wed 24 Jul 2024 09:44:06 AM CEST.
Dependencies resolved.
========================================================================================================================================
 Package                                  Architecture               Version                          Repository                   Size
========================================================================================================================================
Installing:
 zlib-ng-compat-devel                     x86_64                     2.1.7-1.fc40                     updates                      38 k

Transaction Summary
========================================================================================================================================
Install  1 Package

Total download size: 38 k
Installed size: 107 k
Is this ok [y/N]: y
Downloading Packages:
zlib-ng-compat-devel-2.1.7-1.fc40.x86_64.rpm                                                            489 kB/s |  38 kB     00:00
----------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                    62 kB/s |  38 kB     00:00
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                1/1
  Installing       : zlib-ng-compat-devel-2.1.7-1.fc40.x86_64                                                                       1/1
  Running scriptlet: zlib-ng-compat-devel-2.1.7-1.fc40.x86_64                                                                       1/1

Installed:
  zlib-ng-compat-devel-2.1.7-1.fc40.x86_64

Complete!
chipc@fedora:~/workspaces/slowgames/testproj$ dotnet publish -c Release -r linux-x64 -p:PublishAOT=true
MSBuild version 17.8.5+b5265ef37 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  testproj -> /home/chipc/workspaces/slowgames/testproj/bin/Release/net8.0/linux-x64/testproj.dll
  testproj -> /home/chipc/workspaces/slowgames/testproj/bin/Release/net8.0/linux-x64/publish/
```

### Further technical details
```
.NET SDK:
 Version:           8.0.105
 Commit:            eae90abaaf
 Workload version:  8.0.302-baseline.24280.2

Runtime Environment:
 OS Name:     fedora
 OS Version:  40
 OS Platform: Linux
 RID:         fedora.40-x64
 Base Path:   /usr/lib64/dotnet/sdk/8.0.105/

.NET workloads installed:
 Workload version: 8.0.302-baseline.24280.2
There are no installed workloads to display.

Host:
  Version:      8.0.5
  Architecture: x64
  Commit:       087e15321b

.NET SDKs installed:
  8.0.105 [/usr/lib64/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.5 [/usr/lib64/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.5 [/usr/lib64/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/usr/lib64/dotnet]

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

I'm getting restore error on Fedora trying to make simple AOT console project out of the box.

```
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net10.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
        <PublishAot>true</PublishAot>
    </PropertyGroup>

</Project>
```

`error NU1101: Unable to find package runtime.fedora.43-x64.Microsoft.DotNet.ILCompiler.`

`Failed to restore`

For some reason, Fedora sets their RID in a non standard way: `RID: fedora.43-x64` instead of `linux-x64` so the package uri is defined wrong.

`dotnet --info`

```
dotnet --info
.NET SDK:
 Version:           10.0.100
 Commit:            b0f34d51fc
 Workload version:  10.0.100-manifests.4eab3e0f
 MSBuild version:   18.0.2+b0f34d51f

Runtime Environment:
 OS Name:     fedora
 OS Version:  43
 OS Platform: Linux
 RID:         fedora.43-x64
 Base Path:   /usr/lib64/dotnet/sdk/10.0.100/

.NET workloads installed:
There are no installed workloads to display.
Configured to use workload sets when installing new manifests.
No workload sets are installed. Run "dotnet workload restore" to install a workload set.

Host:
  Version:      10.0.0
  Architecture: x64
  Commit:       b0f34d51fc

.NET SDKs installed:
  10.0.100 [/usr/lib64/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 10.0.0 [/usr/lib64/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 10.0.0 [/usr/lib64/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_BUNDLE_EXTRACT_BASE_DIR           [/home/cdock/.cache/dotnet_bundle_extract]
  DOTNET_ROOT                              [/usr/lib64/dotnet]

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download

```

--- a/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/ProcessFrameworkReferencesTests.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/ProcessFrameworkReferencesTests.cs
@@ -82,6 +82,30 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
             }
             """;
 
+        private const string NonPortableRid = "fedora.42-x64";
+
+        private static readonly string NonPortableRuntimeGraph = $$"""
+            {
+                "runtimes": {
+                    "base": {
+                        "#import": []
+                    },
+                    "any": {
+                        "#import": ["base"]
+                    },
+                    "linux": {
+                        "#import": ["any"]
+                    },
+                    "linux-x64": {
+                        "#import": ["linux"]
+                    },
+                    "{{NonPortableRid}}": {
+                        "#import": ["linux-x64"]
+                    }
+                }
+            }
+            """;
+
         // Shared known framework references
         private readonly MockTaskItem _validWindowsSDKKnownFrameworkReference = CreateKnownFrameworkReference(
             "Microsoft.Windows.SDK.NET.Ref", "net5.0-windows10.0.18362", "10.0.18362.1-preview", 
@@ -493,6 +517,7 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
                 ["TargetFramework"] = "net10.0",
                 ["ILCompilerPackNamePattern"] = "runtime.**RID**.Microsoft.DotNet.ILCompiler",
                 ["ILCompilerPackVersion"] = "10.0.0-rc.2.25457.102",
+                ["ILCompilerPortableRuntimeIdentifiers"] = "osx-x64;osx-arm64;win-x64;linux-x64",
                 ["ILCompilerRuntimeIdentifiers"] = "osx-x64;osx-arm64;win-x64;linux-x64"
             });
 
@@ -966,6 +991,53 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
                 task.PackagesToDownload.Should().Contain(p => p.ItemSpec == $"Microsoft.NETCore.App.Runtime.{rid}",
                     $"Should include runtime pack for supported RID: {rid}");
             }
+        }
+
+        [Theory]
+        [InlineData(null, "linux-x64")]
+        [InlineData("linux-x64", "linux-x64")]
+        [InlineData(NonPortableRid, NonPortableRid)]
+        public void It_selects_correct_ILCompiler_based_on_RuntimeIdentifier(string? runtimeIdentifier, string expectedILCompilerRid)
+        {
+            var netCoreAppRef = CreateKnownFrameworkReference("Microsoft.NETCore.App", "net10.0", "10.0.1",
+                "Microsoft.NETCore.App.Runtime.**RID**",
+                "linux-x64;" + NonPortableRid);
+
+            var ilCompilerPack = new MockTaskItem("Microsoft.DotNet.ILCompiler", new Dictionary<string, string>
+            {
+                ["TargetFramework"] = "net10.0",
+                ["ILCompilerPackNamePattern"] = "runtime.**RID**.Microsoft.DotNet.ILCompiler",
+                ["ILCompilerPackVersion"] = "10.0.1",
+                ["ILCompilerPortableRuntimeIdentifiers"] = "linux-x64",
+                ["ILCompilerRuntimeIdentifiers"] = "linux-x64;" + NonPortableRid
+            });
+
+            var config = new TaskConfiguration
+            {
+                TargetFrameworkVersion = "10.0",
+                EnableRuntimePackDownload = true,
+                PublishAot = true,
+                NETCoreSdkRuntimeIdentifier = NonPortableRid,
+                NETCoreSdkPortableRuntimeIdentifier = "linux-x64",
+                RuntimeIdentifier = runtimeIdentifier,
+                RuntimeGraphPath = CreateRuntimeGraphFile(NonPortableRuntimeGraph),
+                FrameworkReferences = new[] { new MockTaskItem("Microsoft.NETCore.App", new Dictionary<string, string>()) },
+                KnownFrameworkReferences = new[] { netCoreAppRef },
+                KnownILCompilerPacks = new[] { ilCompilerPack }
+            };
+
+            var task = CreateTask(config);
+            task.Execute().Should().BeTrue("Task should succeed");
+
+            // Validate that the expected ILCompiler pack is used
+            task.PackagesToDownload.Should().NotBeNull();
+            task.PackagesToDownload.Should().Contain(p => p.ItemSpec.Contains("Microsoft.DotNet.ILCompiler"),
+                "Should include ILCompiler pack when PublishAot is true");
+
+            var ilCompilerPackage = task.PackagesToDownload.FirstOrDefault(p => p.ItemSpec.Contains("Microsoft.DotNet.ILCompiler"));
+            ilCompilerPackage.Should().NotBeNull();
+            ilCompilerPackage!.ItemSpec.Should().Contain(expectedILCompilerRid,
+                $"Should use {expectedILCompilerRid} ILCompiler pack");
         }
     }
 }

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs
@@ -857,7 +857,6 @@ namespace Microsoft.NET.Build.Tasks
                 string? supportedPortableTargetRid = NuGetUtils.GetBestMatchingRid(runtimeGraph, runtimeIdentifier, packSupportedPortableRuntimeIdentifiers, out _);
 
                 bool usePortable = !string.IsNullOrEmpty(NETCoreSdkPortableRuntimeIdentifier)
-                                    && supportedTargetRid is not null && supportedPortableTargetRid is not null
                                     && supportedTargetRid == supportedPortableTargetRid;
 
                 // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture

